components/service/glean/src/main/java/mozilla/components/service/glean/utils/FreshBootDetector.kt

```diff
@@ -0,0 +1,102 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.utils
+
+import android.content.Context
+import android.os.Build
+import android.provider.Settings
+import android.support.annotation.VisibleForTesting
+import java.io.File
+import java.io.IOException
+import java.util.UUID
+
+internal object FreshBootDetector {
+    // cache the value of the first check so we can safely run it multiple times
+    @VisibleForTesting
+    var cacheIsFreshBoot: Boolean? = null
+
+    /**
+     * Gets the boot id, which is a value that changes every time the
+     * device is rebooted.
+     *
+     * For Android Oreo (SDK 24) and later, we can use the [BOOT_COUNT] global
+     * setting.
+     *
+     * For earlier versions of Android, we can use the Linux kernel's
+     * `/proc/sys/kernel/random/boot_id`.
+     *
+     * Adapted from https://github.com/instacart/truetime-android/pull/93
+     *
+     * @param context An application context
+     */
+    internal fun getBootId(context: Context): String {
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
+            try {
+                return Settings.Global.getInt(
+                    context.getContentResolver(),
+                    Settings.Global.BOOT_COUNT
+                ).toString()
+            } catch (e: Settings.SettingNotFoundException) {
+                // pass through to old method
+            }
+        }
+
+        try {
+            val file = File("/proc/sys/kernel/random/boot_id")
+
+            val bootId = StringBuilder()
+
+            file.forEachLine {
+                bootId.append(it)
+            }
+
+            return bootId.toString()
+        } catch (e: IOException) {
+            // pass through to fallback method
+        }
+
+        // If neither of the above methods works, just return
+        // a random id, so it will appear that we are always at
+        // a fresh reboot.
+        return UUID.randomUUID().toString()
+    }
+
+    /**
+     * Returns true if this is the first time the application has run
+     * since the device has rebooted.
+     *
+     * If determining this fails, it returns true by default.
+     *
+     * The first determination is cached in memory, so it is safe to call this
+     * function multiple times.
+     *
+     * @param context An application context
+     * @return true if this is the first time the application has run since
+     *    the device was rebooted.
+     */
+    @Synchronized
+    public fun isFreshBoot(context: Context): Boolean {
+        if (cacheIsFreshBoot != null) {
+            return cacheIsFreshBoot!!
+        }
+
+        val prefs = context.getSharedPreferences(
+            this.javaClass.canonicalName, Context.MODE_PRIVATE
+        )
+
+        val currentBootId = getBootId(context)
+        val storedBootId = prefs.getString("bootId", "")
+
+        cacheIsFreshBoot = (currentBootId != storedBootId)
+
+        if (cacheIsFreshBoot!!) {
```

Comment:
nit: Instead of `!!` it makes more sense to do `!!` on the nullable String.

Or use `== true` or `== false`.